walker_test.go

```diff
@@ -305,3 +304,361 @@ func tmpDir(inp []*change) (dir string, retErr error) {
+		"ADD bar/foo file",
+		"ADD baz dir",
+		"ADD foo2 file",
+		"ADD foo dir",
+		"ADD foo/bar dir",
+		"ADD foo/bar/bee file",
+	}))
+
+	assert.NoError(t, err)
+	defer os.RemoveAll(d)
+	b := &bytes.Buffer{}
+	err = Walk(context.Background(), d, &WalkOpt{
+		IncludePatterns: []string{"**"},
+	}, bufWalk(b))
+	assert.NoError(t, err)
+
+	trimEqual(t, `
+		dir a
+		dir a/b
+		dir a/b/bar
+		file a/b/bar/foo
+		file a/b/bar/fop
+		dir a/b/baz
+		dir bar
+		file bar/foo
+		dir baz
+		dir foo
+		dir foo/bar
+		file foo/bar/bee
+		file foo2
+	`, string(b.Bytes()))
+
+	b.Reset()
+	err = Walk(context.Background(), d, &WalkOpt{
+		IncludePatterns: []string{"**/bar"},
+	}, bufWalk(b))
+	assert.NoError(t, err)
+
+	trimEqual(t, `
+		dir a
+		dir a/b
+		dir a/b/bar
+		file a/b/bar/foo
+		file a/b/bar/fop
+		dir bar
+		file bar/foo
+		dir foo
+		dir foo/bar
+		file foo/bar/bee
+	`, string(b.Bytes()))
+
+	b.Reset()
+	err = Walk(context.Background(), d, &WalkOpt{
+		IncludePatterns: []string{"**/bar/foo"},
+	}, bufWalk(b))
+	assert.NoError(t, err)
+
+	trimEqual(t, `
+		dir a
+		dir a/b
+		dir a/b/bar
+		file a/b/bar/foo
+		dir bar
+		file bar/foo
+	`, string(b.Bytes()))
+
+	b.Reset()
+	err = Walk(context.Background(), d, &WalkOpt{
+		IncludePatterns: []string{"**/b*"},
+	}, bufWalk(b))
+	assert.NoError(t, err)
+
+	trimEqual(t, `
+		dir a
+		dir a/b
+		dir a/b/bar
+		file a/b/bar/foo
+		file a/b/bar/fop
+		dir a/b/baz
+		dir bar
+		file bar/foo
+		dir baz
+		dir foo
+		dir foo/bar
+		file foo/bar/bee
+	`, string(b.Bytes()))
+
+	b.Reset()
+	err = Walk(context.Background(), d, &WalkOpt{
+		IncludePatterns: []string{"**/bar/f*"},
+	}, bufWalk(b))
+	assert.NoError(t, err)
+
+	trimEqual(t, `
+			dir a
+			dir a/b
+			dir a/b/bar
+			file a/b/bar/foo
+			file a/b/bar/fop
+			dir bar
+			file bar/foo
+	`, string(b.Bytes()))
+
+	b.Reset()
+	err = Walk(context.Background(), d, &WalkOpt{
+		IncludePatterns: []string{"**/bar/g*"},
+	}, bufWalk(b))
+	assert.NoError(t, err)
+
+	trimEqual(t, ``, string(b.Bytes()))
+
+	b.Reset()
+	err = Walk(context.Background(), d, &WalkOpt{
+		IncludePatterns: []string{"**/f*"},
+	}, bufWalk(b))
+	assert.NoError(t, err)
+
+	trimEqual(t, `
+		dir a
+		dir a/b
+		dir a/b/bar
+		file a/b/bar/foo
+		file a/b/bar/fop
+		dir bar
+		file bar/foo
+		dir foo
+		dir foo/bar
+		file foo/bar/bee
+		file foo2
+	`, string(b.Bytes()))
+
+	b.Reset()
+	err = Walk(context.Background(), d, &WalkOpt{
+		IncludePatterns: []string{"**/b*/f*"},
+	}, bufWalk(b))
+	assert.NoError(t, err)
+
+	trimEqual(t, `
+		dir a
+		dir a/b
+		dir a/b/bar
+		file a/b/bar/foo
+		file a/b/bar/fop
+		dir bar
+		file bar/foo
+	`, string(b.Bytes()))
+
+	b.Reset()
+	err = Walk(context.Background(), d, &WalkOpt{
+		IncludePatterns: []string{"**/b*/foo"},
+	}, bufWalk(b))
+	assert.NoError(t, err)
+
+	trimEqual(t, `
+		dir a
+		dir a/b
+		dir a/b/bar
+		file a/b/bar/foo
+		dir bar
+		file bar/foo
+	`, string(b.Bytes()))
+
+	// TODO can we support this? Currently filepath.Clean is called
```

Comment:
I think we can replace this comment/block with:
```
	b.Reset()
	err = Walk(context.Background(), d, &WalkOpt{
		IncludePatterns: []string{"**/foo/**"},
	}, bufWalk(b))
	assert.NoError(t, err)

	trimEqual(t, `
		dir foo
		dir foo/bar
		file foo/bar/bee
	`, string(b.Bytes()))
```

I had originally thought `**/foo/` could match all directories named `foo`, but I think `**/foo/**` better represents this idea and seems to work.

Reply:
Updated, thanks.